Added OpInfo-based testing of some linalg functions

IvanYashchuk opened the pull request:
Added OpInfo-based testing of the following linear algebra functions:
* cholesky, linalg.cholesky
* linalg.eigh
* inverse, linalg.inv
* qr, linalg.qr
* solve

The output of `torch.linalg.pinv` for empty inputs was not differentiable, now it's fixed.

In some cases, batched grad checks are disabled because it doesn't work well with 0x0 matrices (see https://github.com/pytorch/pytorch/issues/50743#issuecomment-767376085).

Ref. #50006

IvanYashchuk committed:
Remove unused sample_inputs_pinverse

IvanYashchuk committed:
Use skipCUDAIfNoMagmaAndNoCusolver in OpInfo tests

IvanYashchuk committed:
Added OpInfo for torch.solve

IvanYashchuk committed:
Fixed autograd support of linalg_pinv for empty inputs

IvanYashchuk committed:
Updated pseudo inverse op tests

IvanYashchuk committed:
Added OpInfo entry for torch.inverse

IvanYashchuk committed:
Added OpInfo entry for torch.qr, torch.linalg.qr

IvanYashchuk committed:
Added OpInfo entry for cholesky

IvanYashchuk committed:
Added OpInfo entry for torch.linalg.eigh

IvanYashchuk committed:
Fix magma cusolver decorators

IvanYashchuk commented:
Tests for QR decomposition are quite slow
```
========================================================= slowest 20 durations =========================================================
32.76s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_qr_cuda_complex128
32.08s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_complex128
16.98s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_float64
15.77s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_solve_cuda_complex64
15.09s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_solve_cuda_complex128
14.89s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_float32
13.98s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_solve_cuda_float32
13.53s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_flip_cuda_complex128
13.19s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_complex64
13.05s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_svd_cuda_complex128
13.04s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_svd_cuda_complex128
12.51s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_svd_cuda_float64
11.87s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_complex128
10.38s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_solve_cuda_float32
10.06s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_solve_cuda_complex64
9.71s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_solve_cuda_complex128
9.08s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_qr_cuda_float64
9.05s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_float64
8.80s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_qr_cuda_float64
8.68s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_svd_cuda_float32
===================================== 5139 passed, 1009 skipped, 61 warnings in 1069.53s (0:17:49) =====================================
```

IvanYashchuk committed:
flake8

IvanYashchuk commented:
The failing tests should be fixed after merging https://github.com/pytorch/pytorch/pull/51109.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Added 0x0 case for cholesky inputs

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Merge branch 'master' into linalg-opinfo-tests

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Make lines shorter

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Fix merge; add skipCUDAIfRocm

mruberry commented:
Hey @IvanYashchuk! Just checking-in on this PR. Is it waiting for review?

IvanYashchuk commented:
> Hey @IvanYashchuk! Just checking-in on this PR. Is it waiting for review?

It's ready for review. I was just not pinging you knowing you're busy and hoping that this PR is in your queue :wink:

mruberry commented:
> > Hey @IvanYashchuk! Just checking-in on this PR. Is it waiting for review?
> 
> It's ready for review. I was just not pinging you knowing you're busy and hoping that this PR is in your queue 😉

Haha, in my queue enough to ask you about it. Taking a look now.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1827,13 +1887,26 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
            test_inplace_grad=False,
            supports_tensor_out=False,
-           sample_inputs_func=sample_inputs_slogdet,
+           sample_inputs_func=sample_inputs_linalg_inv,
```
Is there a better name for sample_inputs_linalg_inv if it's being used for slogdet and pinv, too?

IvanYashchuk replied:
I changed it to sample_inputs_linalg_invertible.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1543,6 +1538,21 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+                    sample_inputs_func=sample_inputs_linalg_cholesky,
+                    decorators=[skipCUDAIfNoMagma, skipCUDAIfRocm, skipCPUIfNoLapack],
+                    skips=(
+                        # These tests do not take into account custom op.get_op()
```
This is fine although it'd be nice if HermitianOpInfo applied this skip so each HermitianOpInfo didn't need to.

IvanYashchuk replied:
I agree, I moved this skip to HermitianOpInfo and applied a similar change for TriangularOpInfo.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2089,6 +2173,18 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+           op=torch.solve,
+           dtypes=floating_and_complex_types(),
+           test_inplace_grad=False,
+           # TODO: TypeError: empty_like(): argument 'input' (position 1) must be Tensor, not torch.return_types.solve
```
I appreciate this TODO but this is more a failure of our out= testing (which we should be updating soonish) and not an issue with torch.solve().

mruberry reviewed (approved):
Awesome! Nice work, @IvanYashchuk. I really appreciate how thorough and consistent you were with documentation. It makes the code much more readable and easier to maintain.

This needs a rebase. Just ping me when it's ready to merge.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Renamed sample_inputs_linalg_inv -> invertible

IvanYashchuk committed:
Added default skip to HermitianOpInfo and TriangularOpInfo

IvanYashchuk commented:
@mruberry I updated and rebased this PR. It's ready to merge.

mruberry commented:
The ROCm test failures on this PR are interesting and linalg related. Would you take a look at them before we merge this?

IvanYashchuk committed:
Added skipCUDAIfRocm

IvanYashchuk commented:
It was my fault not being careful enough when resolving the merge conflict and I accidentally removed `skipCUDAIfRocm` for `logdet`. Let's wait for CI now.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Added skipCUDAIfRocm for pseudoinverse tests

ROCm fails with RuntimeError: magma: The value of work_size(-9223372036854775808) is too large to fit into a magma_int_t (4 bytes)

IvanYashchuk commented:
ROCm still fails for 
```
test_variant_consistency_jit_linalg_pinv_cuda_float32
test_variant_consistency_eager_triangular_solve_cuda_float32
test_variant_consistency_eager_linalg_pinv_hermitian_cuda_float32
test_variant_consistency_eager_linalg_pinv_cuda_float32
test_out_linalg_pinv_hermitian_cuda_float32
test_out_linalg_pinv_cuda_float32
```
```
RuntimeError: magma: The value of work_size(-9223372036854775808) is too large to fit into a magma_int_t (4 bytes)

----------------------------------------------------------------------
Ran 9205 tests in 1586.823s

FAILED (errors=6, skipped=2448)
```

IvanYashchuk committed:
More skipCUDAIfRocm

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Merge branch 'master' into linalg-opinfo-tests

IvanYashchuk committed:
Remove supports_tensor_out

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk committed:
Make non-empty sample input to be the first in the list

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-opinfo-tests

IvanYashchuk commented:
@mruberry I updated this pull request with the recent changes for "out" testing. I also added all relevant ROCm skips, which will be fixed sometime later. Could you take a look once again and hopefully merge this?

mruberry commented:
I'll try to get this landed ASAP. I was concerned about the ROCm failure because it's a timeout, but that test build doesn't appear to be running these tests. The overall test time seems similar to current CI timings.

Commit:
Merge branch 'master' into linalg-opinfo-tests
